File: packages/core/src/utils/surql.ts
import type { MutationEventType } from '../types';

// ==================== TYPES ====================

export interface TxQuery {
  readonly __brand: 'TxQuery';
  readonly sql: string;
  readonly statementCount: number;
}

export interface SealOptions {
  /** 0-based index of the inner statement to extract. Default: last statement. */
  resultIndex?: number;
}

export interface SealedQuery<T = void> {
  readonly sql: string;
  readonly extract: (results: unknown[]) => T;
}

export interface SurqlHelper {
  seal(query: string): string;
  seal<T = void>(query: TxQuery, options?: SealOptions): SealedQuery<T>;
  tx(queries: string[]): TxQuery;
  selectById(idVar: string, returnValues: string[]): string;
  selectByFieldsAnd(
    table: string,
    whereVar: ({ field: string; variable: string } | string)[],
    returnValues: ({ field: string; alias: string } | string)[]
  ): string;
  create(idVar: string, dataVar: string): string;
  createSet(
    idVar: string,
    keyDataVars: ({ key: string; variable: string } | { statement: string } | string)[]
  ): string;
  upsert(idVar: string, dataVar: string): string;
  upsertMerge(idVar: string, dataVar: string): string;
  updateMerge(idVar: string, dataVar: string): string;
  updateSet(
    idVar: string,
    keyDataVar: ({ key: string; variable: string } | { statement: string } | string)[]
  ): string;
  delete(idVar: string): string;
  let(name: string, query: string): string;
  createMutation(
    t: MutationEventType,
    mutationIdVar: string,
    recordIdVar: string,
    dataVar?: string,
    beforeRecordVar?: string
  ): string;
  returnObject(entries: { key: string; variable: string }[]): string;
}

// ==================== IMPLEMENTATION ====================

export const surql: SurqlHelper = {
  seal(query: string | TxQuery, options?: SealOptions): any {
    if (typeof query === 'string') {
      return `${query};`;
    }

    // TxQuery path
    const txQuery = query;
    const idx = options?.resultIndex ?? txQuery.statementCount - 1;
    const sql = `${txQuery.sql};`;
    return {
      sql,
      extract(results: unknown[]): unknown {
        // +1 to skip the BEGIN null at index 0
        return results[idx + 1];
      },
    } satisfies SealedQuery<unknown>;
  },

  tx(queries: string[]): TxQuery {
    return {
      __brand: 'TxQuery' as const,
      sql: `BEGIN TRANSACTION;\n${queries.join(';')};\nCOMMIT TRANSACTION`,
      statementCount: queries.length,
    };
  },

  selectById(idVar: string, returnValues: string[]) {
    return `SELECT ${returnValues.join(',')} FROM ONLY $${idVar}`;
  },

  selectByFieldsAnd(
    table: string,
    whereVar: ({ field: string; variable: string } | string)[],
    returnValues: ({ field: string; alias: string } | string)[]
  ) {
    return `SELECT ${returnValues
      .map((rv) => (typeof rv === 'string' ? rv : `${rv.field} as ${rv.alias}`))
      .join(',')} FROM ${table} WHERE ${whereVar
      .map((wv) => (typeof wv === 'string' ? `${wv} = $${wv}` : `${wv.field} = $${wv.variable}`))
      .join(' AND ')}`;
  },

  create(idVar: string, dataVar: string) {
    return `CREATE ONLY $${idVar} CONTENT $${dataVar}`;
  },

  createSet(
    idVar: string,
    keyDataVars: ({ key: string; variable: string } | { statement: string } | string)[]
  ) {
    return `CREATE ONLY $${idVar} SET ${keyDataVars
      .map((keyDataVar) =>
        typeof keyDataVar === 'string'
          ? `${keyDataVar} = $${keyDataVar}`
          : 'statement' in keyDataVar
            ? keyDataVar.statement
            : `${keyDataVar.key} = $${keyDataVar.variable}`
      )
      .join(', ')}`;
  },

  upsert(idVar: string, dataVar: string) {
    return `UPSERT ONLY $${idVar} REPLACE $${dataVar}`;
  },

  // Use this for sync-down ingestion: the remote payload omits local-only
  // fields injected by the CLI's local-schema setup (`_00_crdt`,
  // `_00_cursor`), and REPLACE would wipe them on every round-trip and
  // break offline reload of CRDT state.
  upsertMerge(idVar: string, dataVar: string) {
    return `UPSERT ONLY $${idVar} MERGE $${dataVar}`;
  },

  updateMerge(idVar: string, dataVar: string) {
    return `UPDATE ONLY $${idVar} MERGE $${dataVar}`;
  },

  updateSet(
    idVar: string,
    keyDataVar: ({ key: string; variable: string } | { statement: string } | string)[]
  ) {
    return `UPDATE $${idVar} SET ${keyDataVar
      .map((kdv) =>
        typeof kdv === 'string'
          ? `${kdv} = $${kdv}`
          : 'statement' in kdv
            ? kdv.statement
            : `${kdv.key} = $${kdv.variable}`
      )
      .join(', ')}`;
  },

  delete(idVar: string) {
    return `DELETE $${idVar}`;
  },

  let(name: string, query: string) {
    return `LET $${name} = (${query})`;
  },

  createMutation(
    t: MutationEventType,
    mutationIdVar: string,
    recordIdVar: string,
    dataVar?: string,
    beforeRecordVar?: string
  ) {
    switch (t) {
      case 'create':
        // `data` is REQUIRED here, not decorative. The outbox row is the only
        // copy of a pending create once the in-memory UpEvent is gone (reload,
        // or a shared-tabs follower whose row is replayed by the leader). This
        // used to drop `dataVar` on the floor, so every replayed create arrived
        // with `data: undefined` and `processUpEvent` threw on
        // `Object.keys(event.data)` before it ever reached the network: the
        // create was unsendable AND it sat at the head of the queue blocking
        // every later mutation. The payload was simply never persisted, so
        // those creates were unrecoverable.
        return dataVar
          ? `CREATE ONLY $${mutationIdVar} SET mutationType = 'create', recordId = $${recordIdVar}, data = $${dataVar}`
          : `CREATE ONLY $${mutationIdVar} SET mutationType = 'create', recordId = $${recordIdVar}`;
      case 'update': {
        let stmt = `CREATE ONLY $${mutationIdVar} SET mutationType = 'update', recordId = $${recordIdVar}, data = $${dataVar}`;
        if (beforeRecordVar) {
          stmt += `, beforeRecord = $${beforeRecordVar}`;
        }
        return stmt;
      }
      case 'delete':
        return `CREATE ONLY $${mutationIdVar} SET mutationType = 'delete', recordId = $${recordIdVar}`;
    }
  },

  returnObject(entries: { key: string; variable: string }[]) {
    return `RETURN {${entries.map(({ key, variable }) => `${key}: $${variable}`).join(',')}}`;
  },
};

File: packages/spooky_core/lib/src/utils/surql.dart
import '../types.dart' show MutationEventType;

/// A multi-statement transaction body (TS `TxQuery`).
class TxQuery {
  const TxQuery({required this.sql, required this.statementCount});
  final String sql;
  final int statementCount;
}

/// A query plus a result extractor for one statement of a transaction
/// (TS `SealedQuery<T>`).
class SealedQuery<T> {
  const SealedQuery({required this.sql, required this.extract});
  final String sql;
  final T Function(List<dynamic> results) extract;
}

/// One item of a `SET` / `RETURN` clause. Mirrors the TS union
/// `{ key, variable } | { statement } | string`.
sealed class SetItem {
  const SetItem();

  /// `field = $field` (the bare-string TS branch).
  factory SetItem.field(String field) = _RawField;

  /// `key = $variable`.
  factory SetItem.keyVar(String key, String variable) = _KeyVar;

  /// A raw statement fragment, inserted verbatim.
  factory SetItem.statement(String statement) = _Statement;

  String render();
}

class _RawField extends SetItem {
  const _RawField(this.field);
  final String field;
  @override
  String render() => '$field = \$$field';
}

class _KeyVar extends SetItem {
  const _KeyVar(this.key, this.variable);
  final String key;
  final String variable;
  @override
  String render() => '$key = \$$variable';
}

class _Statement extends SetItem {
  const _Statement(this.statement);
  final String statement;
  @override
  String render() => statement;
}

/// One item of a `WHERE ... AND` clause. Mirrors `{ field, variable } | string`.
sealed class WhereItem {
  const WhereItem();
  factory WhereItem.field(String field) = _RawWhere;
  factory WhereItem.fieldVar(String field, String variable) = _WhereVar;
  String render();
}

class _RawWhere extends WhereItem {
  const _RawWhere(this.field);
  final String field;
  @override
  String render() => '$field = \$$field';
}

class _WhereVar extends WhereItem {
  const _WhereVar(this.field, this.variable);
  final String field;
  final String variable;
  @override
  String render() => '$field = \$$variable';
}

/// One item of a `SELECT` return list. Mirrors `{ field, alias } | string`.
sealed class ReturnItem {
  const ReturnItem();
  factory ReturnItem.raw(String value) = _RawReturn;
  factory ReturnItem.aliased(String field, String alias) = _AliasedReturn;
  String render();
}

class _RawReturn extends ReturnItem {
  const _RawReturn(this.value);
  final String value;
  @override
  String render() => value;
}

class _AliasedReturn extends ReturnItem {
  const _AliasedReturn(this.field, this.alias);
  final String field;
  final String alias;
  @override
  String render() => '$field as $alias';
}

/// SurrealQL builder helpers. Faithful port of the TS `surql` object.
class Surql {
  const Surql();

  /// Seal a single statement string: appends `;`.
  String seal(String query) => '$query;';

  /// Seal a transaction, returning a [SealedQuery] whose [SealedQuery.extract]
  /// reads the chosen inner statement (defaulting to the last). The `+1`
  /// skips the `BEGIN` null at index 0.
  SealedQuery<dynamic> sealTx(TxQuery query, {int? resultIndex}) {
    final idx = resultIndex ?? query.statementCount - 1;
    return SealedQuery<dynamic>(
      sql: '${query.sql};',
      extract: (results) => results[idx + 1],
    );
  }

  /// Build a `BEGIN/COMMIT` transaction wrapping [queries].
  TxQuery tx(List<String> queries) => TxQuery(
        sql: 'BEGIN TRANSACTION;\n${queries.join(';')};\nCOMMIT TRANSACTION',
        statementCount: queries.length,
      );

  String selectById(String idVar, List<String> returnValues) =>
      'SELECT ${returnValues.join(',')} FROM ONLY \$$idVar';

  String selectByFieldsAnd(
    String table,
    List<WhereItem> whereVar,
    List<ReturnItem> returnValues,
  ) =>
      'SELECT ${returnValues.map((rv) => rv.render()).join(',')} '
      'FROM $table WHERE ${whereVar.map((wv) => wv.render()).join(' AND ')}';

  String create(String idVar, String dataVar) =>
      'CREATE ONLY \$$idVar CONTENT \$$dataVar';

  String createSet(String idVar, List<SetItem> keyDataVars) =>
      'CREATE ONLY \$$idVar SET ${keyDataVars.map((k) => k.render()).join(', ')}';

  String upsert(String idVar, String dataVar) =>
      'UPSERT ONLY \$$idVar REPLACE \$$dataVar';

  /// Sync-down ingestion: MERGE (not REPLACE) so local-only fields
  /// (`_00_crdt`, `_00_cursor`) survive the round-trip.
  String upsertMerge(String idVar, String dataVar) =>
      'UPSERT ONLY \$$idVar MERGE \$$dataVar';

  String updateMerge(String idVar, String dataVar) =>
      'UPDATE ONLY \$$idVar MERGE \$$dataVar';

  String updateSet(String idVar, List<SetItem> keyDataVar) =>
      'UPDATE \$$idVar SET ${keyDataVar.map((k) => k.render()).join(', ')}';

  String delete(String idVar) => 'DELETE \$$idVar';

  String let(String name, String query) => 'LET \$$name = ($query)';

  String createMutation(
    MutationEventType t,
    String mutationIdVar,
    String recordIdVar, {
    String? dataVar,
    String? beforeRecordVar,
  }) {
    switch (t) {
      case MutationEventType.create:
        return "CREATE ONLY \$$mutationIdVar SET mutationType = 'create', "
            'recordId = \$$recordIdVar';
      case MutationEventType.update:
        var stmt = "CREATE ONLY \$$mutationIdVar SET mutationType = 'update', "
            'recordId = \$$recordIdVar, data = \$$dataVar';
        if (beforeRecordVar != null) {
          stmt += ', beforeRecord = \$$beforeRecordVar';
        }
        return stmt;
      case MutationEventType.delete:
        return "CREATE ONLY \$$mutationIdVar SET mutationType = 'delete', "
            'recordId = \$$recordIdVar';
    }
  }

  String returnObject(List<(String key, String variable)> entries) =>
      'RETURN {${entries.map((e) => '${e.$1}: \$${e.$2}').join(',')}}';
}

/// Singleton matching the TS `surql` export.
const surql = Surql();
